principles.rst

```diff
@@ -303,3 +309,130 @@ down GHC even when those features are not active. Yet this principle is importan
 as we hope not to make the current situation worse.
 
 From `#378`_, slightly generalized.
+
+
+
+GHC stability principles
+--------------------------
+
+The Haskell ecosystem has a built-in tension between stability and innovation.
+
+One the one hand, breaking changes impose heavy costs:
+
+* Users keep having to update their code
+* Library authors keep having to update their code
+* Updating to a later version of a library can in turn force taking on new dependencies.
+
+These difficulties add friction, discourage adoption, and make Haskell an unpredictable investment.
+
+On the other hand, GHC has always sought to be a laboratory for innovation.  If the stability guarantees are too onerous, we risk imposing a different set of costs:
+
+* Volunteers may get discouraged by the hoops they have to jump through to get a change agreed.
+* The language may become stuck in a local optimum, because moving to a better design would require breaking changes.
+* GHC may become encrusted with cruft and workarounds that exist to preserve existing, but undesirable, behaviour.
+
+We can't navigate this tension with simple all-or-nothing rules.  We have to take each case on its merits, aware of both sets of costs.  Rather than starting the debate from scratch each time, we can have some general rules, or principles, to guide us.  The `GHC base library proposal <https://github.com/haskellfoundation/tech-proposals/blob/main/proposals/accepted/051-ghc-base-libraries.rst>`_ made a start in this direction; this section discusses GHC itself.
+
+Assumptions
+~~~~~~~~~~~~
+
+We assume:
+
+* That we have adopted the  `GHC base library proposal <https://github.com/haskellfoundation/tech-proposals/blob/main/proposals/accepted/051-ghc-base-libraries.rst>`_, which establishes the ``ghc-experimental`` package.
+* That we have identified some extensions as "Experimental".  These extensions may well change, and are subject to
+  much weaker stability goals.  `GHC Proposal #601 <https://github.com/ghc-proposals/ghc-proposals/pull/601>`_, describes a plan along these lines, but all we need for this document is a list of Experimental extensions.
+
+Terminology
+~~~~~~~~~~~~~~~
+
+We define a **stable Haskell package** as follows. A stable Haskell package
+
+* Does not use Experimental extensions.
+* Does not use experimental features.  (Examples: use of the Javascript or Wasm back end, builds on non-tier-1 platforms.)
+* Does not rely on explicitly-undefined behaviour. (Example: ``INCOHERENT`` instance selection.)
+* Does not use ``-Werror`` in its default build configuration.
+* Explicitly specifies a language edition (``Haskell98``, ``GHC2021``), in the source code or the build configuration.
+
+Stability (GR1)
+~~~~~~~~~~~~~~~~
+
```

Comment:
(GR1) says (roughly) that if a package compiles with a particular GHC-X then it should compile with a later GHC-Y, where Y>X.   In principle it would be good if it also compiled with an *earlier* GHC-Z, where Z<X.   Of course, at a minimum, GHC-Z should understand all the extensions that GHC-X does.

If we extend extensions, even in non-breaking ways, that would cease to be true.  But it would mainly cease to be true in cases where GHC-Z just wouldn't understand the new code anyway, so it hardly matters.  

Example, suppose that instead of extending `ExplicitNameSpaces` we instead added a new extension `ExtendedExplicitNameSpaces` to cover the delta.  Well, the same packages P that need `ExtendedExplicitNameSpaces` then of course won't compile with GHC-Z (because it lacks that extension). 

So if we add a new extension
* We add some extra bureaucracy in documentation (we have a lot of extensions) and implementation.
* We don't gain much: the difference is that when compiling pacakge P with GHC-Z, we get "Undefined extension `ExtendedExplicitNameSpaces`" rather than a parse error.

Personally I think the pain does not justify the gain.  So I think that, in general, I'm OK with adding non-breaking functionality to an extension.   What do others think?   I think it would help if the final proposal was explicit on this point.

Reply:
E.g. `other-extensions` mentioned in https://github.com/ghc-proposals/ghc-proposals/issues/623#issuecomment-1837589501, see https://cabal.readthedocs.io/en/stable/cabal-package.html#pkg-field-other-extensions

Reply:
> Personally I think the pain does not justify the gain. So I think that, in general, I'm OK with adding non-breaking functionality to an extension. What do others think?

This reminds me of the idea of "language edition as first class" that we scratched in https://github.com/ghc-proposals/ghc-proposals/pull/620

If language editions were first class, functionality changes would only be permitted in a new edition. 

Language extensions would then simply become partial configuration of the language edition. And the language edition would potentially be more than the sum of language extension knobs.

My feeling is that I'd rather not have extensions evolve, but the language/features somehow must. So until we get more agreement on the "first class language edition" idea, there's no better alternative than to allow adding non-breaking changes (we want less extensions, not more, right?).